common/src/main/java/io/pravega/common/util/btree/sets/BTreeSet.java

```diff
@@ -0,0 +1,414 @@
+        // to ensure that parent pages aren't updated before leaf pages (which would cause index corruptions - i.e., by
+        // pointing to inexistent pages).
+        val processedPageIds = new HashSet<Long>();
+
+        // First collect updates. Begin from the bottom (Leaf Pages).
+        val toWrite = new ArrayList<Map.Entry<Long, ArrayView>>();
+        collectWriteCandidates(pageCollection.getLeafPages(), toWrite, processedPageIds, pageCollection);
+
+        // Newly split pages may not be reachable from any modified Leaf Pages. Collect them too.
+        collectWriteCandidates(pageCollection.getIndexPages(), toWrite, processedPageIds, pageCollection);
+
+        // Then collect deletions, making sure we also consider all their parents (which should be modified/deleted as well).
+        collectWriteCandidates(pageCollection.getDeletedPagesParents(), toWrite, processedPageIds, pageCollection);
+        log.debug("{}: Persist (Updates={}, Deletions={}).", this.traceLogId, toWrite.size(), pageCollection.getDeletedPageIds().size());
+        return this.update.apply(toWrite, pageCollection.getDeletedPageIds(), timer.getRemaining());
+    }
+
+    private void collectWriteCandidates(Collection<BTreeSetPage> candidates, List<Map.Entry<Long, ArrayView>> toWrite,
+                                        Set<Long> processedIds, PageCollection pageCollection) {
+        while (!candidates.isEmpty()) {
+            val next = new ArrayList<BTreeSetPage>();
+            candidates.stream()
+                    .filter(p -> p.isModified() && !processedIds.contains(p.getPagePointer().getPageId()))
+                    .forEach(p -> {
+                        toWrite.add(new AbstractMap.SimpleImmutableEntry<>(p.getPagePointer().getPageId(), p.getData()));
+                        val parent = pageCollection.get(p.getPagePointer().getParentPageId());
+                        assert p.getPagePointer().hasParent() == (parent != null);
+                        processedIds.add(p.getPagePointer().getPageId());
+                        if (parent != null) {
+                            next.add(parent);
+                        }
+                    });
+            candidates = next;
+        }
+    }
+
+    //endregion
+
+    //region Queries
+
+    /**
+     * Returns an {@link AsyncIterator} that will iterate through all the items in this {@link BTreeSet} within the
+     * specified bounds. All iterated items will be returned in lexicographic order (smallest to largest).
+     * See {@link ByteArrayComparator} for ordering details.
+     *
+     * @param firstItem          An {@link ArrayView} indicating the first Item to iterate from. If null, the iteration
+     *                           will begin with the first item in the index.
+     * @param firstItemInclusive If true, firstIem will be included in the iteration (provided it exists), otherwise it
```

Comment:
I'm curious, why you think it is a good idea to have booleans for inclusion rather than pick one option and go with it?

Reply:
Because sometimes I need to iterate starting with an element and sometimes I need to iterate immediately after an element (i.e., I know the last element I processed, so give me the next one).

Reply:
> if the caller has the first item and wants it inclusive, then it can use the first item it already has.

Reply:
> The caller may have the first item, but the first item may not be part of the set. If we always make it exclusive, the caller may erroneously include that first item in the result even if it is not part of the set.

It is correct that making it always exclusive requires a different code path to add the first or last when you want it inclusive, and that makes it more prone to errors. The developer can always set the flag incorrectly, though.

> This is a low-level API.

This is part of common, which is principle should be usable even outside Pravega. I'm curious, is this common in the implementation of iterators?  Do you know of other examples outside Pravega that implement iterators like this?

> Another reason would be to make it consistent

We should make it consistent, at least for this PR as we are not really changing `BTreeIndex`.

Reply:
Some examples I could quickly came up with:
- Java `IntStream.range(start, end)` and `IntStream.rangeClosed(start, end)`
- Java `NavigableSet`: [headSet](https://docs.oracle.com/javase/7/docs/api/java/util/NavigableSet.html#headSet(E,%20boolean) and [subSet](https://docs.oracle.com/javase/7/docs/api/java/util/NavigableSet.html#subSet(E,%20boolean,%20E,%20boolean))
    - Similarly for `NavigableMap`

NavigableSet and NavigableMap are what I based the APIs for BTreeSet and BTreeIndex on (more or less).